File: src/dexter/frontend/clang/src/ir/array/IncrPtrExpr.cpp
 /*
 * IncrPtrExpr.cpp
 *
 *  Created on: Mar 24, 2018
 *      Author: maazsaf
 */

#include "ir/array/IncrPtrExpr.h"

Dexter::IncrPtrExpr::IncrPtrExpr (Dexter::Expr * data, Dexter::Expr * offset) : Dexter::Expr(NULL)
{
	JNIEnv * e = Util::env;
  jclass c = e->FindClass("dexter/ir/array/IncrPtrExpr");
	jmethodID m = e->GetMethodID(c, "<init>", "(Ldexter/ir/Expr;Ldexter/ir/Expr;)V");
	this->mObj = e->NewObject(c, m, data->obj(), offset->obj());
}

Dexter::Expr * Dexter::IncrPtrExpr::ptr ()
{
  JNIEnv * e = Util::env;
  jclass c = e->FindClass("dexter/ir/array/IncrPtrExpr");
  jmethodID m = e->GetMethodID(c, "ptr", "()Ldexter/ir/Expr;");
  jobject id = e->CallObjectMethod(this->mObj, m);
  return (id == NULL) ? NULL : new Dexter::Expr(id);
}

Dexter::Expr * Dexter::IncrPtrExpr::incr ()
{
  JNIEnv * e = Util::env;
  jclass c = e->FindClass("dexter/ir/array/IncrPtrExpr");
  jmethodID m = e->GetMethodID(c, "incr", "()Ldexter/ir/Expr;");
  jobject id = e->CallObjectMethod(this->mObj, m);
  return (id == NULL) ? NULL : new Dexter::Expr(id);
}

File: src/dexter/ir/array/IncrPtrExpr.java
package dexter.ir.array;

import dexter.ir.Expr;
import dexter.ir.Visitor;
import dexter.ir.bool.CallExpr;

import java.util.Arrays;

/**
 * Created by Maaz Ahmad on 6/25/19.
 */
public class IncrPtrExpr extends CallExpr
{
  public IncrPtrExpr(Expr ptr, Expr incr)
  {
    super("incr", Arrays.asList(ptr, incr));
  }

  public Expr ptr () { return this.args.get(0); }
  public Expr incr  () { return this.args.get(1); }

  @Override public <T> T accept(Visitor<T> p) { return p.visit(this); }
}